File: connectors/whatsapp-web/src/media-storage.ts
/**
 * S3-compatible uploader for WhatsApp media. New writes go to the shared
 * skirmshop-drive bucket; raw legacy keys are still read from the old MinIO.
 * Object keys follow `socialmedia/attachments/{messageId}/{ts}.{ext}`.
 *
 * On startup the connector calls `ensureMediaBucket()` to make idempotent the bucket
 * existence check (cheap — single statObject call).
 */
import * as MinIO from 'minio';
import * as fs from 'fs';
import * as https from 'https';

const MINIO_ENDPOINT = process.env.S3_ENDPOINT || process.env.MINIO_ENDPOINT || 'minio:9000';
const MINIO_ACCESS_KEY =
  process.env.AWS_ACCESS_KEY_ID || process.env.MINIO_ACCESS_KEY || 'minioadmin';
const MINIO_SECRET_KEY =
  process.env.AWS_SECRET_ACCESS_KEY || process.env.MINIO_SECRET_KEY || 'minioadmin';
const MINIO_USE_SSL =
  (process.env.S3_USE_SSL || process.env.MINIO_USE_SSL || 'true').toLowerCase() === 'true';
const MINIO_CA_CERT = process.env.MINIO_CA_CERT;
const BUCKET = process.env.S3_BUCKET || process.env.MINIO_BUCKET || 'socialmedia-media';
const PREFIX = (process.env.S3_PREFIX || '').replace(/^\/+|\/+$/g, '');
const LEGACY_MINIO_ENDPOINT =
  process.env.LEGACY_MINIO_ENDPOINT || process.env.MINIO_ENDPOINT || MINIO_ENDPOINT;
const LEGACY_MINIO_ACCESS_KEY =
  process.env.LEGACY_MINIO_ACCESS_KEY || process.env.MINIO_ACCESS_KEY || MINIO_ACCESS_KEY;
const LEGACY_MINIO_SECRET_KEY =
  process.env.LEGACY_MINIO_SECRET_KEY || process.env.MINIO_SECRET_KEY || MINIO_SECRET_KEY;
const LEGACY_MINIO_USE_SSL =
  (process.env.LEGACY_MINIO_USE_SSL || process.env.MINIO_USE_SSL || 'true').toLowerCase() ===
  'true';
const LEGACY_BUCKET =
  process.env.LEGACY_MINIO_BUCKET || process.env.MINIO_BUCKET || 'socialmedia-media';
// Out-of-cluster S3 endpoint (e.g. the Traefik LAN IngressRoute
// https://skirmshop-s3.lan.e-dani.com). Used ONLY to presign GET URLs that
// consumers outside the cluster (synapse adapters on sauvage) can fetch —
// presigned URLs embed the signed host, so the in-cluster svc endpoint would
// produce URLs that never resolve out there. Unset ⇒ presigning disabled.
const S3_PUBLIC_ENDPOINT = process.env.S3_PUBLIC_ENDPOINT || '';
const S3_PRESIGN_EXPIRY_SECONDS = parseInt(process.env.S3_PRESIGN_EXPIRY_SECONDS || '3600', 10);

function endpointParts(
  endpoint: string,
  fallbackSsl: boolean
): { host: string; port: number; useSSL: boolean } {
  const withScheme = /^https?:\/\//i.test(endpoint)
    ? endpoint
    : `${fallbackSsl ? 'https' : 'http'}://${endpoint}`;
  const parsed = new URL(withScheme);
  return {
    host: parsed.hostname,
    port: parsed.port ? parseInt(parsed.port, 10) : parsed.protocol === 'https:' ? 443 : 9000,
    useSSL: parsed.protocol === 'https:',
  };
}

function clientFor(
  endpoint: string,
  accessKey: string,
  secretKey: string,
  useSSL: boolean
): MinIO.Client {
  const parts = endpointParts(endpoint, useSSL);
  const opts: MinIO.ClientOptions = {
    endPoint: parts.host,
    port: parts.port,
    useSSL: parts.useSSL,
    accessKey,
    secretKey,
  };
  if (parts.useSSL && MINIO_CA_CERT && fs.existsSync(MINIO_CA_CERT)) {
    const ca = fs.readFileSync(MINIO_CA_CERT, 'utf-8');
    opts.transportAgent = new https.Agent({ ca, rejectUnauthorized: true });
  }
  return new MinIO.Client(opts);
}

const client = clientFor(MINIO_ENDPOINT, MINIO_ACCESS_KEY, MINIO_SECRET_KEY, MINIO_USE_SSL);
const legacyClient = clientFor(
  LEGACY_MINIO_ENDPOINT,
  LEGACY_MINIO_ACCESS_KEY,
  LEGACY_MINIO_SECRET_KEY,
  LEGACY_MINIO_USE_SSL
);
// Same credentials as `client`, different (publicly reachable) signed host.
const publicClient = S3_PUBLIC_ENDPOINT
  ? clientFor(S3_PUBLIC_ENDPOINT, MINIO_ACCESS_KEY, MINIO_SECRET_KEY, MINIO_USE_SSL)
  : null;

function withPrefix(key: string): string {
  const clean = key.replace(/^\/+/, '');
  return PREFIX ? `${PREFIX}/${clean}` : clean;
}

export function parseStorageRef(ref: string): { bucket: string; key: string; legacy: boolean } {
  if (ref.startsWith('s3://')) {
    const parsed = new URL(ref);
    return { bucket: parsed.hostname, key: parsed.pathname.slice(1), legacy: false };
  }
  return { bucket: LEGACY_BUCKET, key: ref, legacy: true };
}

function s3Uri(key: string): string {
  return `s3://${BUCKET}/${key}`;
}

const EXT_BY_MIME: Record<string, string> = {
  'image/jpeg': 'jpg',
  'image/png': 'png',
  'image/gif': 'gif',
  'image/webp': 'webp',
  'video/mp4': 'mp4',
  'video/quicktime': 'mov',
  'video/webm': 'webm',
  'audio/mpeg': 'mp3',
  'audio/ogg': 'ogg',
  'audio/wav': 'wav',
  'audio/mp4': 'm4a',
  'application/pdf': 'pdf',
};

function pickExt(mimeType: string | undefined, fileName: string | undefined): string {
  if (fileName && fileName.includes('.')) {
    return fileName.split('.').pop()!.slice(0, 6);
  }
  if (mimeType && EXT_BY_MIME[mimeType]) return EXT_BY_MIME[mimeType];
  if (mimeType?.startsWith('image/')) return 'bin';
  return 'bin';
}

export async function ensureMediaBucket(): Promise<void> {
  const exists = await client.bucketExists(BUCKET).catch(() => false);
  if (!exists) {
    await client.makeBucket(BUCKET, 'us-east-1');
  }
}

export interface UploadedMedia {
  storageKey: string;
  fileSize: number;
}

export async function uploadMedia(
  messageId: bigint | number,
  data: Buffer,
  mimeType: string | undefined,
  fileName: string | undefined
): Promise<UploadedMedia> {
  const ts = Date.now();
  const ext = pickExt(mimeType, fileName);
  const storageKey = withPrefix(`attachments/${messageId}/${ts}.${ext}`);
  const meta: Record<string, string> = {};
  if (mimeType) meta['Content-Type'] = mimeType;
  if (fileName) meta['x-amz-meta-filename'] = fileName;
  await client.putObject(BUCKET, storageKey, data, data.length, meta);
  return { storageKey: s3Uri(storageKey), fileSize: data.length };
}

/** Configured presign expiry (seconds); ridden as attachment metadata. */
export function presignExpirySeconds(): number {
  return S3_PRESIGN_EXPIRY_SECONDS;
}

/**
 * Presigned GET URL for a stored object, signed against S3_PUBLIC_ENDPOINT so
 * consumers OUTSIDE the cluster (synapse adapters on sauvage) can fetch it
 * without credentials or headers (auth rides in the query string).
 *
 * Returns null when presigning is unavailable: no public endpoint configured,
 * or the ref points at the legacy MinIO (which has no out-of-cluster
 * exposure). Presigning is a local signature computation — no network call.
 */
export async function presignMediaUrl(
  storageKey: string,
  expirySeconds: number = S3_PRESIGN_EXPIRY_SECONDS
): Promise<string | null> {
  if (!publicClient) return null;
  const ref = parseStorageRef(storageKey);
  if (ref.legacy) return null;
  return publicClient.presignedGetObject(ref.bucket, ref.key, expirySeconds);
}

/** Fetch a stored object from MinIO by its storage key, as a Buffer. */
export async function fetchMedia(storageKey: string): Promise<Buffer> {
  const ref = parseStorageRef(storageKey);
  const source = ref.legacy ? legacyClient : client;
  const stream = await source.getObject(ref.bucket, ref.key);
  const chunks: Buffer[] = [];
  for await (const chunk of stream) chunks.push(chunk as Buffer);
  return Buffer.concat(chunks);
}

/**
 * Upload a profile picture under avatars/{kind}/{base64url(id)}.jpg.
 * Returns the storage key so the caller can save it into
 * conversations.avatar_url / participants.profile_pic_url.
 */
export async function uploadAvatar(
  kind: 'conversations' | 'participants',
  ident: string,
  data: Buffer
): Promise<string> {
  await ensureMediaBucket();
  const safe = Buffer.from(ident).toString('base64url');
  const storageKey = withPrefix(`avatars/${kind}/${safe}.jpg`);
  await client.putObject(BUCKET, storageKey, data, data.length, {
    'Content-Type': 'image/jpeg',
  });
  return s3Uri(storageKey);
}

File: connectors/telegram-sync/sync/media_storage.py
"""MinIO uploader for Telegram media — used by history.py + realtime.py + backfill.

Bucket `socialmedia-media` is shared across WA + TG + IG. Object keys follow
`attachments/{message_id}/{ts}.{ext}` so they don't collide. We upload bytes
directly from BytesIO (no disk write).
"""
from __future__ import annotations

import io
import logging
import os
import time
from typing import Optional

from minio import Minio

logger = logging.getLogger(__name__)

MINIO_ENDPOINT = os.environ.get("S3_ENDPOINT") or os.environ.get("MINIO_ENDPOINT", "minio:9000")
MINIO_ACCESS_KEY = os.environ.get("AWS_ACCESS_KEY_ID") or os.environ.get("MINIO_ACCESS_KEY", "minioadmin")
MINIO_SECRET_KEY = os.environ.get("AWS_SECRET_ACCESS_KEY") or os.environ.get("MINIO_SECRET_KEY", "minioadmin")
MINIO_USE_SSL = os.environ.get("S3_USE_SSL", os.environ.get("MINIO_USE_SSL", "true")).lower() == "true"
MINIO_BUCKET = os.environ.get("S3_BUCKET") or os.environ.get("MINIO_BUCKET", "socialmedia-media")
S3_PREFIX = os.environ.get("S3_PREFIX", "").strip("/")
LEGACY_MINIO_ENDPOINT = os.environ.get("LEGACY_MINIO_ENDPOINT") or os.environ.get("MINIO_ENDPOINT", MINIO_ENDPOINT)
LEGACY_MINIO_ACCESS_KEY = os.environ.get("LEGACY_MINIO_ACCESS_KEY") or os.environ.get("MINIO_ACCESS_KEY", MINIO_ACCESS_KEY)
LEGACY_MINIO_SECRET_KEY = os.environ.get("LEGACY_MINIO_SECRET_KEY") or os.environ.get("MINIO_SECRET_KEY", MINIO_SECRET_KEY)
LEGACY_MINIO_USE_SSL = os.environ.get("LEGACY_MINIO_USE_SSL", os.environ.get("MINIO_USE_SSL", "true")).lower() == "true"
LEGACY_MINIO_BUCKET = os.environ.get("LEGACY_MINIO_BUCKET") or os.environ.get("MINIO_BUCKET", "socialmedia-media")
# Use cert_check=False to accept the cluster's self-signed CA. The CA file is
# at /certs/ca.crt inside the container; for a stricter setup use
# `Minio(..., http_client=urllib3.PoolManager(ca_certs="/certs/ca.crt"))`.
_CERT_CHECK = os.environ.get("MINIO_CERT_CHECK", "false").lower() == "true"

_client: Optional[Minio] = None
_legacy_client: Optional[Minio] = None


def _endpoint(endpoint: str) -> str:
    return endpoint.removeprefix("http://").removeprefix("https://").rstrip("/")


def _get_client() -> Minio:
    global _client
    if _client is None:
        _client = Minio(
            _endpoint(MINIO_ENDPOINT),
            access_key=MINIO_ACCESS_KEY,
            secret_key=MINIO_SECRET_KEY,
            secure=MINIO_USE_SSL or MINIO_ENDPOINT.startswith("https://"),
            cert_check=_CERT_CHECK,
        )
    return _client


def _get_legacy_client() -> Minio:
    global _legacy_client
    if _legacy_client is None:
        _legacy_client = Minio(
            _endpoint(LEGACY_MINIO_ENDPOINT),
            access_key=LEGACY_MINIO_ACCESS_KEY,
            secret_key=LEGACY_MINIO_SECRET_KEY,
            secure=LEGACY_MINIO_USE_SSL or LEGACY_MINIO_ENDPOINT.startswith("https://"),
            cert_check=_CERT_CHECK,
        )
    return _legacy_client


def _with_prefix(key: str) -> str:
    key = key.lstrip("/")
    return f"{S3_PREFIX}/{key}" if S3_PREFIX else key


def _s3_uri(key: str) -> str:
    return f"s3://{MINIO_BUCKET}/{key}"


def parse_storage_ref(ref: str) -> tuple[str, str, bool]:
    if ref.startswith("s3://"):
        from urllib.parse import urlparse
        parsed = urlparse(ref)
        return parsed.netloc, parsed.path.lstrip("/"), False
    return LEGACY_MINIO_BUCKET, ref, True


def ensure_bucket() -> None:
    c = _get_client()
    if not c.bucket_exists(MINIO_BUCKET):
        c.make_bucket(MINIO_BUCKET)
        logger.info(f"Created MinIO bucket: {MINIO_BUCKET}")


_EXT_BY_MIME = {
    "image/jpeg": "jpg",
    "image/png": "png",
    "image/gif": "gif",
    "image/webp": "webp",
    "video/mp4": "mp4",
    "video/quicktime": "mov",
    "audio/ogg": "ogg",
    "audio/mpeg": "mp3",
    "audio/mp4": "m4a",
    "audio/wav": "wav",
    "application/pdf": "pdf",
}


def pick_ext(mime_type: Optional[str], file_name: Optional[str]) -> str:
    if file_name and "." in file_name:
        return file_name.rsplit(".", 1)[-1][:6].lower()
    if mime_type and mime_type in _EXT_BY_MIME:
        return _EXT_BY_MIME[mime_type]
    if mime_type and mime_type.startswith("image/"):
        return "bin"
    return "bin"


def upload_media(
    message_id: int,
    data: bytes,
    mime_type: Optional[str],
    file_name: Optional[str],
) -> tuple[str, int]:
    """Upload bytes to MinIO and return (storage_key, size). Storage key goes
    into attachments.file_url; the dashboard streams via /api/messages/media-blob/{id}."""
    c = _get_client()
    ext = pick_ext(mime_type, file_name)
    storage_key = _with_prefix(f"attachments/{message_id}/{int(time.time() * 1000)}.{ext}")
    buf = io.BytesIO(data)
    metadata = {}
    if mime_type:
        metadata["Content-Type"] = mime_type
    c.put_object(
        MINIO_BUCKET,
        storage_key,
        buf,
        length=len(data),
        content_type=mime_type or "application/octet-stream",
        metadata=metadata or None,
    )
    return _s3_uri(storage_key), len(data)


def download_media(storage_ref: str) -> bytes:
    bucket, key, legacy = parse_storage_ref(storage_ref)
    response = (_get_legacy_client() if legacy else _get_client()).get_object(bucket, key)
    try:
        return response.read()
    finally:
        response.close()
        response.release_conn()


def upload_avatar(kind: str, ident: str, data: bytes) -> str:
    """Upload a profile picture to MinIO under avatars/{kind}/{base64url(id)}.jpg.

    kind ∈ {conversations, participants}. The base64url(id) encoding handles
    WhatsApp `@`-suffixed JIDs and Telegram negative IDs safely.
    """
    import base64 as _b64
    c = _get_client()
    ensure_bucket()
    safe = _b64.urlsafe_b64encode(ident.encode()).decode().rstrip("=")
    storage_key = _with_prefix(f"avatars/{kind}/{safe}.jpg")
    buf = io.BytesIO(data)
    c.put_object(
        MINIO_BUCKET,
        storage_key,
        buf,
        length=len(data),
        content_type="image/jpeg",
    )
    return _s3_uri(storage_key)
